packages/backend/src/invoice/service.ts

```diff
@@ -100,6 +123,39 @@ async function createInvoice(
   }
 }
 
+// Deactivate expired invoices that have some money.
+// Delete expired invoices that have never received money.
+// Returns the id of the processed invoice (if any).
+async function deactivateNextInvoice(
+  deps: ServiceDependencies
+): Promise<string | undefined> {
+  return deps.knex.transaction(async (trx) => {
+    // 30 seconds backwards to allow a prepared (but not yet fulfilled/rejected) packet to finish before being deactivated.
+    const now = new Date(Date.now() - 30_000).toISOString()
+    const invoices = await Invoice.query(trx)
+      .limit(1)
+      // Ensure the invoices cannot be processed concurrently by multiple workers.
+      .forUpdate()
+      // If an invoice is locked, don't wait — just come back for it later.
+      .skipLocked()
+      .where('active', true)
+      .andWhere('expiresAt', '<', now)
+    const invoice = invoices[0]
+    if (!invoice) return
+
+    // Fetch the account with a separate transaction to avoid locking it.
```

Comment:
What's the downside of locking the account? The only thing we'd currently patch on it is the `disabled` field.

Reply:
You mean locking instead of deleting it? The endpoint to create invoices is unauthenticated, so there's concern about spam (see: #123).

Reply:
Nevermind, `locking` here refers to Postgres.